File: WebKitPluginView.rb
# Superclass NSTextField is already defined in WebKitPluginView.h
class WebKitPluginView
  def initWithArguments(attributes)
    if initWithFrame(NSZeroRect)
      puts "Initialized with attributes: #{attributes.description}"

      # This is how you access the enclosing WebView instance
      @webView = attributes[WebPlugInContainerKey].webFrame.webView

      # This makes the textfield call evalRubyCode when the user hits enter in the textfield
      self.target = self
      self.action = "evalRubyCode:"

      self
    end
  end

  # sender is actually self in this case...
  def evalRubyCode(sender)
    eval sender.stringValue

    # Notify the WebView's UIDelegate, normally you'd have to check if the UIDelegate actually
    # responds to the message, but in this case it is known that the TestBrowser's AppController
    # instance does in fact do so
    if @webView.UIDelegate.respond_to?("webView:textField:didReceiveText:")
      @webView.UIDelegate.webView(@webView, textField: self, didReceiveText: self.stringValue)
    end
  end

  # WebPlugIn informal protocol
  #
  # All these methods are not required and can safely be removed.

  # This method will be only called once per instance of the plug-in object, and will be called
  # before any other methods in the WebPlugIn protocol.
  def webPlugInInitialize
    puts "[!] webPlugInInitialize called"
  end

  # The plug-in usually begins drawing, playing sounds and/or animation in this method.
  def webPlugInStart
    puts "[!] webPlugInStart called"
  end

  # The plug-in normally stop animations/sounds in this method.
  def webPlugInStop
    puts "[!] webPlugInStop called"
  end

  # Perform cleanup and prepare to be deallocated.
  def webPlugInDestroy
    puts "[!] webPlugInDestroy called"
  end

  # This is typically used to allow the plug-in to alter its appearance when selected
  def webPlugInSetIsSelected(selected)
    puts "[!] webPlugInSetIsSelected called: #{selected}"
  end

  # Returns the object that exposes the plug-in's interface.  The class of this object can implement
  # methods from the WebScripting informal protocol.
  def objectForWebScript
    puts "[!] objectForWebScript called"
    self
  end

end

File: WebKitPluginView.m
#import "WebKitPluginView.h"
#import <MacRuby/MacRuby.h>

//@interface WebKitPluginView (Internal)
//- (id)_initWithArguments:(NSDictionary *)arguments;
//@end

@implementation WebKitPluginView

// WebPlugInViewFactory protocol
// The principal class of the plug-in bundle must implement this protocol.

+ (NSView *)plugInViewWithArguments:(NSDictionary *)newArguments
{
  NSBundle *bundle = [NSBundle bundleForClass:[WebKitPluginView class]];
  NSURL *url = [bundle URLForResource:@"WebKitPluginView" withExtension:@"rb"];
  [[MacRuby sharedRuntime] evaluateFileAtURL:url];
  return [[self alloc] performSelector:@selector(initWithArguments:)];
}

// WebPlugIn informal protocol

//- (void)webPlugInInitialize
//{
    //// This method will be only called once per instance of the plug-in object, and will be called
    //// before any other methods in the WebPlugIn protocol.
    //// You are not required to implement this method.  It may safely be removed.
//}

//- (void)webPlugInStart
//{
    //// The plug-in usually begins drawing, playing sounds and/or animation in this method.
    //// You are not required to implement this method.  It may safely be removed.
//}

//- (void)webPlugInStop
//{
    //// The plug-in normally stop animations/sounds in this method.
    //// You are not required to implement this method.  It may safely be removed.
//}

//- (void)webPlugInDestroy
//{
    //// Perform cleanup and prepare to be deallocated.
    //// You are not required to implement this method.  It may safely be removed.
//}

//- (void)webPlugInSetIsSelected:(BOOL)isSelected
//{
    //// This is typically used to allow the plug-in to alter its appearance when selected.
    //// You are not required to implement this method.  It may safely be removed.
//}

//- (id)objectForWebScript
//{
    //// Returns the object that exposes the plug-in's interface.  The class of this object can implement
    //// methods from the WebScripting informal protocol.
    //// You are not required to implement this method.  It may safely be removed.
    //return self;
//}

@end

//@implementation WebKitPluginView (Internal)

//- (id)_initWithArguments:(NSDictionary *)newArguments
//{
    //if (!(self = [super initWithFrame:NSZeroRect]))
        //return nil;
    
    //return self;
//}

//@end
